nixos/github-runner: add PAT support

veehaitch opened the pull request:
###### Description of changes

This commit introduces support for runner registrations through a
personal access token (PAT). To use a PAT instead of a registration
token, place an appropriately scoped PAT in `tokenFile`. If the file
contains a PAT, the module queries a new runner registration token from
GitHub. Only github.com is support (i.e., no enterprise setups). Using a
runner registration token directly continues to work as before.

I decided to go for a Python script (without third-party dependencies) as a Bash script was too fragile and cumbersome.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

kfollesdal reviewed (approved):
Nice improvement 👍 Look good 😄

kfollesdal commented:
Result of `nixpkgs-review pr 181823` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

newAM commented on nixos/modules/services/continuous-integration/github-runner/query_registration_token.py:
```diff
@@ -0,0 +1,85 @@
+import sys, json
```
```suggestion
import json
```

`sys` import is unused.

newAM commented on nixos/modules/services/continuous-integration/github-runner/query_registration_token.py:
```diff
@@ -0,0 +1,85 @@
+    try:
+        with urlopen(req) as resp:
+            return json.load(resp)["token"]
+    except HTTPError as err:
+        raise RuntimeError(
+            "Caught unexpected status {err.status} while getting registration token:\n{err.read().decode()}"
+        )
```
```suggestion
    except HTTPError as err:
        raise RuntimeError(
            "Caught unexpected status {err.status} while getting registration token:\n{err.read().decode()}"
        ) from err
```

This is just a cosmetic change for the exception traceback, it makes it clear that the `RuntimeError` exception occurred as the result of the `HTTPError`, instead of the default which (incorrectly) indicates that the `RuntimeError` occured while trying to fix the `HTTPError`.

newAM reviewed (approved):
Have not tested it, but the code looks good.

Thanks for adding this feature, I will definitely be using it!

SuperSandro2000 commented on nixos/modules/module-list.nix:
```diff
@@ -338,7 +338,7 @@
   ./services/continuous-integration/hercules-ci-agent/default.nix
   ./services/continuous-integration/hydra/default.nix
-  ./services/continuous-integration/github-runner.nix
+  ./services/continuous-integration/github-runner/default.nix
```
default.nix can be dropped

SuperSandro2000 commented on nixos/modules/services/continuous-integration/github-runner/default.nix:
```diff
@@ -213,7 +217,10 @@ in
                 echo "Configuring GitHub Actions Runner"
 
-                token=$(< "$STATE_DIRECTORY"/${newConfigTokenFilename})
+                token=$(${pkgs.python3}/bin/python3 ${./query_registration_token.py} \
```
Use python.interpreter to get the path to the python binary.

SuperSandro2000 commented on nixos/modules/services/continuous-integration/github-runner/query_registration_token.py:
```diff
@@ -0,0 +1,85 @@
+
+
+if __name__ == "__main__":
+    import argparse
```
Imports belong to the top of python files.

veehaitch committed:
nixos/github-runner: add PAT support

This commit introduces support for runner registrations through a
personal access token (PAT). To use a PAT instead of a registration
token, place an appropriately scoped PAT in `tokenFile`. If the file
contains a PAT, the module queries a new runner registration token from
GitHub. Only github.com is support (i.e., no enterprise setups). Using a
runner registration token directly continues to work as before.

veehaitch commented:
While browsing the runner source code, I noticed that the configuration routine has PAT support built in which does pretty much the same thing like the Python script in this PR—it just isn't mentioned in the docs. Stay tuned for a follow-up PR.
